---
title: "RNA_seq_DEA"
output: html_document
date: "2025-05-16"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
#1 Install R package, RStudio, directory set-up 
#2 GitHub
#3 Assignment


#4 Count normalization
##4.2 Count normalization of our dataset with DESeq2
###4.2.1 Install packages
Install packages if needed.
```{r, message=FALSE, warning=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2",force = TRUE)
BiocManager::install("edgeR",force = TRUE)
```
###4.2.2 Build a file description
Build or modify your file description to correspond to the files needed and their metadata.
###4.2.3 Read data files

```{r, message=FALSE, warning =FALSE}
Coinfection.targets<-read.delim("./data/fileDesc.txt")
rownames(Coinfection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")

library(edgeR)
library(DESeq2)
```
Here we can check the data file dimension. It has 6 indivuals, for 51751 tags
```{r}
Coinfection.orig <- readDGE(Coinfection.targets, header=F)
dim(Coinfection.orig)
```
To see the 6 first rows of the data, the function head() is used.
```{r}
head(Coinfection.orig)
```
Then build the Extact counts dataframe

```{r}
Coinfection.rawCount <- Coinfection.orig$count
dim(Coinfection.rawCount)
head(Coinfection.rawCount)
```


###4.2.4 Build the metadata

We have 2 categoria of samples : Ha-infection and control
To define them as such in r, we use the function factor()

```{r}
sampletype <- factor(c(rep("Ha",3), rep("Ctr", 3)))
```

metadata frame:

```{r}
meta <- data.frame(sampletype, row.names = colnames(Coinfection.orig$count))
```
Check names of columns and rows and if that sample names match in both files
```{r}
colnames(Coinfection.orig$count)
rownames(meta)
all(colnames(Coinfection.orig$count) %in% rownames(meta))
```

###4.2.5 Create DESeqDataset object

```{r}
library(DESeq2)
dds <- DESeqDataSetFromMatrix(Coinfection.orig, colData = meta, design = ~ sampletype)
head(counts(dds))
```

###4.2.6 Generate size factors

To perform the median of ratios method of normalization, DESeq2 has a single estimateSizeFactors() function that will generate size factors.
normalization factor applied to each samples
```{r}
dds <- estimateSizeFactors(dds)
sizeFactors(dds)
```

###4.2.7 Generate the normalized counts
Now, to retrieve the normalized counts matrix from dds, we use the counts() function and add the argument normalized=TRUE.
Save.

```{r}
normalized_counts <- counts(dds, normalized=TRUE)
write.csv(normalized_counts, file="./results/coinfection_normalized_counts_DESeq2.csv")
```

#5 Sample-level quality Control
##5.1 Principal Component Analysis (PCA)
###5.1.1 PCA overview

To explore the similarity of our samples, we will be performing sample-level QC using Principal Component Analysis (PCA) and hierarchical clustering methods. These methods/tools allow us to check how well similar the replicates are to each other (clustering) and to make sure that the experimental condition is the major source of variation in the data. Sample-level QC can also help identify any samples behaving like outliers; we can further explore any potential outliers to determine whether they need to be removed prior to DE analysis.
he log2 transformation improves the distances/clustering for visualization. Instead of using an ordinary log2 transform, we will be using regularized log transform (rlog), to avoid any bias from the abundance of low-count genes;

###5.1.2 Using our dataset for PCA
####Transform normalized counts for the dataset

Transform with rlog() transformation the normalized counts for data visualization.
```{r}
rld <- rlog(dds, blind=TRUE)
```

The blind=TRUE argument is to make sure that the rlog() function does not take our sample groups into account - i.e. does the transformation in an unbiased manner. When performing quality assessment, it is important to include this option. The DESeq2 vignette has more details about this.

The rlog() function returns a DESeqTransform object, another type of DESeq-specific object. The reason you don’t just get a matrix of transformed values is because all of the parameters (i.e. size factors) that went into computing the rlog transform are stored in that object. We use this object to plot the PCA and heirarchical clustering figures for quality assessment.
####PCA
Using plotPCA() function, which as 2 inputs :
- DESeqTransform object
- “intgroup” (interesting group), which correspond to the column in metadata that has information about the experimental sample groups.
```{r}
# for a simple plot :
DataPCA<-plotPCA(rld, intgroup="sampletype", returnData=TRUE)
DataPCA$sample <- rownames(DataPCA)
percentVar <- round(100 * attr(DataPCA, "percentVar"))
# to have a more ergonomic plot :

library(ggplot2)
library(ggrepel)
pcaData <- plotPCA(rld, intgroup="sampletype", returnData=TRUE)
pcaData$sample <- rownames(pcaData)
percentVar <- round(100 * attr(pcaData, "percentVar"))

p <- ggplot(pcaData, aes(PC1, PC2, color=sampletype, label=sample)) +
  geom_point(size=3) +
  geom_text_repel() +  #Optional: for better label spacing
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5)) +  #Center the title
  ggtitle("PCA plot") 

print(p)
dev.off()

```
Exercise:
It is possible to discern 2 groups on the PCA plot: the control group and the one with Ha inoculation, which have grown less than Control group.The results coincides with bibliography, as *Heterobasidion annousm* is a pathogen of Scots pine.
The first and second axis explain 75% of the variance information, and indicate a high level of data quality.


```{r}
pdf("./results/PlotPCA_dds.pdf")
plotPCA(rld, intgroup="sampletype")
dev.off()
```


#5.2 Hierarchical Clustering Heatmap
Extract the data, compute the pairwise correlation values and then check the head of the data. 
```{r}
rld_mat <- assay(rld)
rld_cor <- cor(rld_mat) 
head(rld_cor)
```
Plot heatmap using the correlation matrix and the metadata object
```{r}
head(meta)
library(pheatmap)
pheatmap(rld_cor, annotation = meta)
```


#6.Differential expression analysis (DEA) using EdgeR
##6.1 Read the individual file by a meta file called fileDesc.txt
Check Coinfection.targets, change the rawnames of the dataframe Coinfection.targets, and read and merges a set of text files containing gene expression counts (makes a DGEList object directly)
```{r}
infection.targets<-read.delim("./data/fileDesc.txt")
infection.targets
library(edgeR)
rownames(infection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")
infection <- readDGE(infection.targets, header=F)
```

##6.2 Raw count distribution
Get the raw mapped count 
```{r}
infection.rawCount <- infection$count
head(infection.rawCount)
library(ggplot2)
```

histogram of the counts for a single sample
xlim and ylim are used to scale the graph.
```{r}
ggplot(infection.rawCount) +
  geom_histogram(aes(x = Ha1), stat = "bin", bins = 200) +
  xlim(0,1500) +
  ylim(0,5000) +
  xlab("Raw expression counts") +
  ylab("Number of genes")
```
There are 51 000 genes, most of them (40 000) have zero values, which means they are not found. 5000 genes are near zero, and only a few have higher expression count, around a hundred.

##6.3 Count normalization with TMM in EdgeR

Scale the library size =Reads of the samples
```{r}
write.csv(infection.rawCount, file="./results/infection.rawCounts.csv")
infection.normCPM <- cpm(calcNormFactors(infection))
dim(infection.normCPM)
head(infection.normCPM)
```

```{r}
write.csv(infection.normCPM, file="./results/infection.normCPM.csv")
```



##6.4 Filter counts per million (CPM) out and get smaller size of libraries
Filter genes that less thn 1 CPM in at least 3 libraries
```{r}
infection.filtered <- rowSums(cpm(infection)>1) >=3
table(infection.filtered)
infection$samples$lib.size
```
cover the original file with filtered data
```{r}
Infection <- infection[infection.filtered,]
colSums(Infection$counts)
dim(Infection)
```
Update the filtered libraries size
```{r}
Infection$samples$lib.size <- colSums(Infection$counts)
Infection$samples
```

##6.5 Count normalization after filter

Performed normalisation with TMM method
```{r}
Infection = calcNormFactors(Infection)
Infection$samples
Infection.filtered.normCPM <-cpm(calcNormFactors(Infection))
write.csv(Infection.filtered.normCPM, file="./results/Infection.filtered.normCPM.csv")
```


##6.6 Experimental design
Design matrix. 
```{r}
group<-factor(c('Ha','Ha','Ha',"Ctr","Ctr","Ctr"))
Infection.design <- model.matrix(~group)   
rownames(Infection.design)<-colnames(Infection$counts)
Infection.design
```
##6.7 Sample-level quality control by multidimensional scaling (MDS)
to see if some samples are outliner. An MDS plot shows the relative similarities of the six samples.
```{r}
plotMDS(Infection, main="MDS plot of RNA-Seq", labels=colnames(Infection$counts))
```

  There is a clear separation between the "Ha" samples and the "Ctr" samples along the first dimension (dim 1). This separation suggests differential gene expression between the two groups, as seen earlier. The relatively high percentage of variance captured (35% + 25% = 60%) indicates that these two dimensions provide a good summary of the major differences.

##6.8 Estimating the common dispersion, trended dispersion, tagwwise dispersion

```{r, warning=FALSE}
Infection <- estimateGLMCommonDisp(Infection, Infection.design)
Infection <- estimateGLMTrendedDisp(Infection, Infection.design)
Infection <- estimateGLMTagwiseDisp(Infection, Infection.design)

plotMeanVar(Infection, show.tagwise.vars=T,NBline=T)
```
```{r}
plotBCV(Infection)
```


##6.9 Fit DGEList and design matrix to genewise GLM

Linear model
coef 2 bc  we use the 2d factor as control

```{r}
Infection.fit <- glmFit(Infection, Infection.design)
colnames(Infection.fit)
lrt.Ha_vs_Ctr <- glmLRT(Infection.fit, coef=2)  # to compare Ha vs Ctr (Ha_vs_Ctr)
```

```{r}
t1<-topTags(lrt.Ha_vs_Ctr, n=nrow(Infection))
head(t1$table)

```


##6.10 Extract number of differentially expressed (DE) genes

```{r}
summary(decideTests(lrt.Ha_vs_Ctr, adjust.method="BH", p.value=0.05))
```

UP regulated genes
```{r}
nrow(subset(topTags(lrt.Ha_vs_Ctr, n=586)$table,  logFC > 0))
lrt.Ha_vs_Ctr_UP <- subset(topTags(lrt.Ha_vs_Ctr, n=586)$table, logFC > 0)
```

Down regulated genes
```{r}
nrow(subset(topTags(lrt.Ha_vs_Ctr, n=586)$table,  logFC < 0))
lrt.Ha_vs_Ctr_DW <- subset(topTags(lrt.Ha_vs_Ctr, n=586)$table, logFC < 0)
```

##6.11 Differentially expressed transcripts’ tag


Export list of UP-regulated and DW-regulated transcripts
```{r}
DEtags.lrt.Ha_vs_Ctr <- rownames(Infection)[as.logical(decideTests(lrt.Ha_vs_Ctr, adjust.method="BH", p.value=0.05))]
write.csv(lrt.Ha_vs_Ctr_UP, file="./results/lrt.Ha_vs_Ctr_UP.csv")
write.csv(lrt.Ha_vs_Ctr_DW, file="./results/lrt.Ha_vs_Ctr_DW.csv")
```

Set all genes as grey color 
```{r}
Infection.colHavsCtr = rep('grey55', nrow(Infection))
```

And assign colors for interesting DE transcripts
```{r}
Infection.colHavsCtr[lrt.Ha_vs_Ctr$table$PValue < 0.05 & lrt.Ha_vs_Ctr$table$logFC >0 ] <- "red"
Infection.colHavsCtr[lrt.Ha_vs_Ctr$table$PValue < 0.05 & lrt.Ha_vs_Ctr$table$logFC <0 ] <- "blue"
```


Plot:

```{r}
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, de.tags=DEtags.lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", pch=19, cex=0.4, smearWidth=0.5, panel.first=grid(), smooth.scatter=FALSE, ylim=c(-7,7), yaxs="i")

abline(h=c(-1,1),col="dodgerblue")
```
All dots are a gene. In gray dots are represented the gene that doesn't have a difference of expression between Ha group and Control. In red are the up-regulated gene of Ha inoculation compared to the control, while in blue are represented the down-regulated genes.
```{r}
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")

abline(h=c(-1,1),col="dodgerblue")
```





